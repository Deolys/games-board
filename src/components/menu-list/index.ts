export {MenuListComposition} from './menu-list';
