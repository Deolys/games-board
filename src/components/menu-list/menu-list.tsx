import { Button, ClickAwayListener, Grow, MenuItem, MenuList, Paper, Popper } from "@mui/material";
import { useEffect, useRef, useState } from "react";
import { Link } from "react-router-dom";

export function MenuListComposition() {
  const [open, setOpen] = useState(false);
  const anchorRef = useRef<HTMLButtonElement>(null);

  const handleToggle = () => {
    setOpen((prevOpen) => !prevOpen);
  };

  const handleClose = (event: Event | React.SyntheticEvent) => {
    if (
      anchorRef.current &&
      anchorRef.current.contains(event.target as HTMLElement)
    ) {
      return;
    }

    setOpen(false);
  };

  const prevOpen = useRef(open);
  useEffect(() => {
    if (prevOpen.current === true && open === false) {
      anchorRef.current!.focus();
    }

    prevOpen.current = open;
  }, [open]);

  return (
    <>
        <Button
          ref={anchorRef}
          id="composition-button"
          aria-controls={open ? 'composition-menu' : undefined}
          aria-expanded={open ? 'true' : undefined}
          aria-haspopup="true"
          onClick={handleToggle}
          sx={{color: "#151724", fontWeight: 700, fontSize: 16}}
        >
          Games
        </Button>
        <Popper
          open={open}
          anchorEl={anchorRef.current}
          role={undefined}
          placement="bottom-start"
          transition
          disablePortal
        >
          {({ TransitionProps, placement }) => (
            <Grow
              {...TransitionProps}
              style={{
                transformOrigin:
                  placement === 'bottom-start' ? 'left top' : 'left bottom',
              }}
            >
              <Paper>
                <ClickAwayListener onClickAway={handleClose}>
                  <MenuList
                    autoFocusItem={open}
                    id="composition-menu"
                    aria-labelledby="composition-button"
                  >
                    <Link to="/rock-paper-scissors">
                    <MenuItem onClick={handleClose}>
                      Rock-Paper-Scissors
                    </MenuItem>
                    </Link>
                    <Link to="/snake">
                    <MenuItem onClick={handleClose}>
                      Snake
                    </MenuItem>
                    </Link>
                    <Link to="/simon-says">
                    <MenuItem onClick={handleClose}>
                      Simon Says
                    </MenuItem>
                    </Link>
                    <Link to="/tic-tac-toe">
                    <MenuItem onClick={handleClose}>
                      Tic-Tac-Toe
                    </MenuItem>
                    </Link>
                  </MenuList>
                </ClickAwayListener>
              </Paper>
            </Grow>
          )}
        </Popper>
      </>
  );
}

export default MenuListComposition;