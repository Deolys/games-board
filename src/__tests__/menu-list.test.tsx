import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { MenuListComposition } from '../components/menu-list';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom';

describe('MenuListComposition', () => {
  it('MenuListComposition mounts properly', () => {
    const { container } = render(<MenuListComposition />);
    expect(container).toMatchSnapshot();

  });

  it('menu opens and closes on button click', async () => {
    render(
      <BrowserRouter>
        <MenuListComposition />
      </BrowserRouter>
    );

    // Проверяем, что меню закрыто изначально
    expect(screen.queryByText('Rock-Paper-Scissors')).not.toBeInTheDocument();

    // Кликаем по кнопке для открытия меню
    fireEvent.click(screen.getByText('Games'));
    expect(screen.queryByText('Rock-Paper-Scissors')).toBeInTheDocument();

    // Кликаем снова, чтобы закрыть меню
    fireEvent.click(screen.getByText('Games'));
    await waitFor(() => {
      expect(screen.queryByText('Rock-Paper-Scissors')).not.toBeInTheDocument();
    });
  });

  it('menu closes when an item is clicked', async () => {
    render(
      <BrowserRouter>
        <MenuListComposition />
      </BrowserRouter>
    );

    // Открываем меню
    fireEvent.click(screen.getByText('Games'));
    expect(screen.queryByText('Rock-Paper-Scissors')).toBeInTheDocument();

    // Кликаем по элементу меню
    fireEvent.click(screen.getByText('Rock-Paper-Scissors'));
    await waitFor(() => {
      expect(screen.queryByText('Rock-Paper-Scissors')).not.toBeInTheDocument();
    });
  });
});

