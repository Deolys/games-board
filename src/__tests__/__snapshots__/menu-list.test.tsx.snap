// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuListComposition MenuListComposition mounts properly 1`] = `
<div>
  <button
    aria-haspopup="true"
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-17ovnxk-MuiButtonBase-root-MuiButton-root"
    id="composition-button"
    tabindex="0"
    type="button"
  >
    Games
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
